FiducialRegistrationWizard/Logic/vtkPointMatcher.cxx

```diff
@@ -0,0 +1,368 @@
+    for ( int pointIndex = 0; pointIndex < numberOfPoints; pointIndex++ )
+    {
+      int permutedPointIndex = pointSubset2IndexPermutations[ permutationIndex][ pointIndex ];
+      double* permutedPoint = pointSubset2->GetPoint( permutedPointIndex );
+      permutedPointSubset2->SetPoint( pointIndex, permutedPoint );
+    }
+    permutedPointSubset2DistanceMatrix->SetPointList1( permutedPointSubset2 );
+    permutedPointSubset2DistanceMatrix->SetPointList2( permutedPointSubset2 );
+    permutedPointSubset2DistanceMatrix->Update();
+    double rootMeanSquareDistanceErrorMm = this->ComputeRootMeanSquareistanceErrors( pointSubset1DistanceMatrix, permutedPointSubset2DistanceMatrix );
+
+    // TODO
+    // if suitability within some threshold of the best
+    //   flag ambiguous
+
+    if ( rootMeanSquareDistanceErrorMm < this->ComputedRootMeanSquareDistanceErrorMm )
+    {
+      this->ComputedRootMeanSquareDistanceErrorMm = rootMeanSquareDistanceErrorMm;
+      this->OutputPointList1->DeepCopy( pointSubset1 );
+      this->OutputPointList2->DeepCopy( permutedPointSubset2 );
+    }
+    
+    // TODO
+    // if new best is outside threshold of the current best
+    //   flag NOT ambiguous
+  }
+}
+
+//------------------------------------------------------------------------------
+double vtkPointMatcher::ComputeRootMeanSquareistanceErrors( vtkPointDistanceMatrix* distanceMatrix1, vtkPointDistanceMatrix* distanceMatrix2 )
+{
+  if ( distanceMatrix1 == NULL || distanceMatrix2 == NULL )
+  {
+    vtkGenericWarningMacro( "One of the input distance matrices is null. Cannot compute similarity. Returning default value " << RESET_VALUE_COMPUTED_ROOT_MEAN_DISTANCE_ERROR << "." );
+    return RESET_VALUE_COMPUTED_ROOT_MEAN_DISTANCE_ERROR;
+  }
+
+  vtkSmartPointer< vtkDoubleArray > distanceErrorMatrix = vtkSmartPointer< vtkDoubleArray >::New();
+  vtkPointDistanceMatrix::ComputePairWiseDifferences( distanceMatrix2, distanceMatrix1, distanceErrorMatrix );
+
+  double sumOfSquaredDistanceErrors = 0;
+  int numberOfColumns = distanceErrorMatrix->GetNumberOfTuples();
+  int numberOfRows = distanceErrorMatrix->GetNumberOfComponents();
+  for ( int columnIndex = 0; columnIndex < numberOfColumns; columnIndex++ )
+  {
+    for ( int rowIndex = 0; rowIndex < numberOfRows; rowIndex++ )
+    {
+      double currentDistanceError = distanceErrorMatrix->GetComponent( columnIndex, rowIndex );
+      sumOfSquaredDistanceErrors += ( currentDistanceError * currentDistanceError );
+    }
+  }
+
+  int numberOfDistances = numberOfColumns * numberOfRows;
+  double meanOfSquaredDistanceErrors = sumOfSquaredDistanceErrors / numberOfDistances;
+  double rootMeanSquareistanceErrors = sqrt( meanOfSquaredDistanceErrors );
+
+  return rootMeanSquareistanceErrors;
+}
+
+//------------------------------------------------------------------------------
+bool vtkPointMatcher::UpdateNeeded()
+{
+  return ( this->InputChangedTime > this->OutputChangedTime );
```

Comment:
If you use standard VTK get/set methods, those update ModifiedTime of the object, so you don't need to maintain a separate InputChangedTime variable.